sdk/communication/Azure.Communication.Administration/src/Models/InternalPhoneNumberReservationOperation.cs

```diff
@@ -0,0 +1,135 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core;
+using Azure.Core.Pipeline;
+
+namespace Azure.Communication.Administration.Models
+{
+    internal class InternalPhoneNumberReservationOperation
+    {
+        private readonly object _lockObject = new object();
+        private readonly PhoneNumberAdministrationClient _client;
+        private readonly CancellationToken _cancellationToken;
+        private bool _hasCompleted;
+        private PhoneNumberReservation? _value;
+        private Response _rawResponse;
+        private Response _finalRawResponse;
+        private readonly IReadOnlyList<ReservationStatus> _terminateStatuses;
+
+        /// <summary>
+        /// Initializes a new <see cref="PhoneNumberReservationOperation"/> instance
+        /// </summary>
+        /// <param name="client">The client used to check for completion.</param>
+        /// <param name="phoneNumberReservationId">The reservation ID of this operation.</param>
+        /// <param name="initialResponse">The original server response on start operation request.</param>
+        /// <param name="terminateStatuses">The list of <see cref="ReservationStatus"/> that indicate the end of operation.</param>
+        /// <param name="cancellationToken">The cancellation token to use.</param>
+        internal InternalPhoneNumberReservationOperation(
+            PhoneNumberAdministrationClient client,
+            string phoneNumberReservationId,
+            Response initialResponse,
+            IReadOnlyList<ReservationStatus> terminateStatuses,
```

Comment:
See comment at the bottom of this file, I'm afraid we need to distinguish success and failure terminate statuses.